R/tableofdifferences.R

```diff
@@ -163,75 +162,61 @@ TableOfDifferences <- function(table1,
         stop("The primary statistic of Table 1 (", stat1[1],
         ") does not match the primary statistic of Table 2 (",
         stat2[1], ").")
+
+    allowed.primary <- c("Average", "%", "Column %", "Row %", "Total %")
+    if (!(stat1[1] %in% allowed.primary))
+        stop("The primary statistic in the input table cannot be '", stat1[1],
+             "'. A t-test can only be computed on '",
+             paste(allowed.primary, collapse = "', '"), "'.")
+
     if (stat1[1] == "Column %")
     {
         n.stat.name <- c("Column Sample Size", "Column n")
         se.stat.name <- "Column Standard Error"
 
     } else if (stat1[1] == "Row %")
     {
-        n.stat.name <- c("Row Sample Size", "Row n")
-        se.stat.name <- "Row Standard Error"
-
+        n.stat.name <- if (q.type == "PickOneMulti") c("Sample Size", "Base n")
+                       else                         c("Row Sample Size", "Row n")
+        se.stat.name <- if (q.type == "PickOneMulti") "Standard Error"
+                        else                         "Row Standard Error" # does not exist
     } else
     {
         n.stat.name <- c("Sample Size", "Base n")
         se.stat.name <- "Standard Error"
     }
+
+    # Get sample size and standard errors
+    # If appropriate SEs are not found, try to recompute using the sample size
     ind1.n <- findIndexOfStat(stat1, n.stat.name)
     ind2.n <- findIndexOfStat(stat2, n.stat.name)
     ind1.se <- findIndexOfStat(stat1, se.stat.name)
     ind2.se <- findIndexOfStat(stat2, se.stat.name)
-
-    if (length(ind1.n) == 0 || #length(ind1.se) == 0 ||
-        length(ind2.n) == 0 ) #|| length(ind2.se) == 0)
+    se1 <- if (length(ind1.se) > 0) table1[,,ind1.se]
+    se2 <- if (length(ind2.se) > 0) table2[,,ind2.se]
+    .calcSE <- function(p, n) return(sqrt(p * (1-p)/n * n/(n-1)))
+    pct.stats <- c("Row %", "Columm %", "Total %", "%")
+    if (is.null(se1) && length(ind1.n) > 0 && stat1[1] %in% pct.stats)
+        se1 <- .calcSE(table1[,,1]/100, table1[,,ind1.n])
+    if (is.null(se2) && length(ind2.n) > 0 && stat2[1] %in% pct.stats)
+        se2 <- .calcSE(table2[,,1]/100, table1[,,ind2.n])
+    if (length(ind1.n) == 0 || is.null(se1) || length(ind2.n) == 0 || is.null(se2))
```

Comment:
The readability of the code could be improved by assigning the conditional in each of these three if statements to a variable with an informative name describing what is being checked.

Reply:
This code is old, but there is still a line in the most recent version checking if any of the sample size/standard error statistics are missing. I deliberately made the error message specify everything that is required, because its annoying to get a message about "Missing Sample Size" and the when you've added the sample size to get another error message about "Missing Standard Error".